desc/objectives/objective_funs.py

```diff
@@ -170,16 +177,17 @@ def build(self, eq, use_jit=None, verbose=1):
         # build objectives
         self._dim_f = 0
         for objective in self.objectives:
-            if verbose > 0:
-                print("Building objective: " + objective.name)
-            objective.build(eq, use_jit=self.use_jit, verbose=verbose)
+            if not objective.built:
```

Comment:
If the equilibrium changes resolution, but the same objective is used, with this conditional here the objective won't be rebuilt, even if one explicitly did 
```
eq = Equilibrium(L=3,M=3)
f_obj = ForceBalance()
obj = ObjectiveFunction(f_obj)
eq.solve(maxiter=5,objective=obj)
eq.change_resolution(L=5,M=5)
obj.build(eq) # it is already built, so this won't do anything
eq.solve(maxiter=5,objective=obj) # fails with a mismatch in shapes
```
I guess one could re-create the `ObjectiveFunction` entirely, but even then, if it is built with previously-created `Objective` objects, then this will not correctly re-build those objects.

I guess my gripe here is that having this here means one MUST re-create new `Objective` objects in order to change the resolution, which might not be a problem but I think if one changes resolution and then rebuilds an `ObjectiveFunction` with that eq, that should be enough to have that `ObjectiveFunction` be built with the correct resolutions

Reply:
The motivation for this was that the user will often call `objective.build` themselves before passing it to the optimizer, and the optimizer would then build again. Similarly, in the course of a regular solve/optimization/perturb run we end up rebuilding stuff way more often than we need to so was trying to cut down on that.

I think this has been a general problem for a while, and I'm not sure what the best solution is. Do we want to always rebuild the objective, even if it's not needed? Or do we try to detect when things change, which might not always be easy?

I'm open to suggestions.

Reply:
Could we make it more robust by adding a try/except block to the line that would fail in the common cases? like for this case:
```python
eq = Equilibrium(L=3,M=3)
f_obj = ForceBalance()
obj = ObjectiveFunction(f_obj)
eq.solve(maxiter=5,objective=obj)


eq.change_resolution(L=5,M=5)
obj.build(eq)
eq.solve(maxiter=5,objective=obj)
```
the error is in line `165         x0 = objective.x(eq)` in desc/optimize/optimizer.py and is a `ValueError`

and for here
```python
eq = Equilibrium(L=3,M=3)
f_obj = ForceBalance()
obj = ObjectiveFunction(f_obj)
eq.solve(maxiter=5,objective=obj)
eq.change_resolution(L=5,M=5)
obj = ObjectiveFunction(f_obj)
obj.build(eq)
eq.solve(maxiter=5,objective=obj)
```

it is line `153             objective.compile(mode, verbose)` of desc/optimize/optimizer.py and is also a `ValueError`

something like
```python
            try:
                objective.compile(mode, verbose)
            except ValueError:
                objective.build(eq, verbose=verbose)
                objective.compile(mode, verbose)
```

I know this may not seem clean though but it would alleviate the double build issue you mention while also addressing the possible confusion the above cases could cause.

Though for this to actually work, we would need to add a `force_rebuild=False` flag to ObjectiveFunction so that if you call `ObjectiveFunction.build(eq,force_rebuild=True)`, it will rebuild every `Objective` it contains even if they are built already.

My vote would be in favor of this approach, or if we don't want to have these try/except blocks, then still including the `force_rebuild` flag with a default `False` and adding something in the description of `ObjectiveFunction` that if the eq resolution is changed one must rebuild with force_rebuild, or recreate the underlying objectives.

Reply:
I think a better option would be to somehow automatically check if something needs to be built within the build method, so you could call it multiple times and have it do nothing if not needed. We could probably do something like cache the resolution properties of the eq when its first built and then compare that to a new eq if you call build again. 

Tricky thing might be profiles and grids. We should go over it at the meeting later.

Reply:
OK for now I went with the try/catch since that covers most of the use cases and is fairly simple. Also added tests for the above use cases.

Eventually we should come up with a better option though.